src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -216,6 +365,208 @@ DropOrphanedShards(bool waitForLocks)
 }
 
 
+/*
+ * StartNewOperationNeedingCleanup is be called by an operation to register
+ * for cleanup.
+ */
+OperationId
+StartNewOperationNeedingCleanup(void)
+{
+	CurrentOperationId = GetNextOperationId();
+
+	LockOperationId(CurrentOperationId);
+
+	return CurrentOperationId;
+}
+
+
+/*
+ * CompleteNewOperationNeedingCleanup is be called by an operation to signal
+ * completion. This will trigger cleanup of appropriate resources.
+ */
+void
+CompleteNewOperationNeedingCleanup(bool isSuccess)
```

Comment:
Complete as a mirror of Start is a bit strange to me, since we are actually still continuing the operation, since the transaction can still fail and we still hold the lock. We could do more stuff that generates additional records.

Maybe something along the lines of CleanCurrentObjects. I'm also wondering whether it would help to have two functions (for failure and success), to not have to explain as many thing.

Reply:
How about I call it 'PrepareForCompleteNewOperationNeedingCleanup' ? The contract here is that this API is to be called just before operation is about to complete. The user should not do anything that generates additional records after that. I could add some extra checks for that?

Reply:
> PrepareForCompleteNewOperationNeedingCleanup

I would prefer function names that state what they do, e.g. DoCleanupForCurrentOperation

> I like the fact that for a Start we have a Complete pair, makes it more elegant IMO.

I think it would be elegant if it matched reality. Our operation ends exactly when the transaction ends, not before. This is true for practically everything in postgres. Perhaps Start is not the right name?

Reply:
@marcocitus 
**The thing I don't like about 'DoCleanupForCurrentOperation' is this :**

1.  The name 'DoCleanupForCurrentOperation' kind of gives a false impression that we only need to call this in case of failure. As a new reader of code, why would you want to clean up in success? (unless you know the internals)

2. It sort of gives an impression that are we are deleting a resource which is not always necessarily true.
There are scenarios where we will not clean any resource but just clean up the pg_dist_cleanup records. For example : The split children record need cleanup on success.

Another possible name for pair : 
RegisterNewOperationNeedingCleanup (instead of start), and FinalizeNewOperationNeedingCleanup (instead of complete) ?

Reply:
Register+Finalize seems nicer than Start+Complete to me

> The name 'DoCleanupForCurrentOperation' kind of gives a false impression that we only need to call this in case of failure.

I think splitting into two functions could be nice.

> 'DoCleanupForCurrentOperation' sort of gives an impression that are we are deleting a resource which is not always necessarily true. Again in success scenario we will not clean any resource but just clean up the pg_dist_cleanup records. For example : The split children record need cleanup on success.

Cleanup does not necessarily imply deleting resource to me. However, I agree that success vs. failure difference could be made more explicit, even if we go for Finalize

Reply:
Updated to RegisterNewOperationNeedingCleanup and FinalizeNewOperationNeedingCleanupOnSuccess / FinalizeNewOperationNeedingCleanupOnFailure.